sdk/tests/performance/webgl_webcodecs_video_frame/webgl_webcodecs_video_frame.js

```diff
@@ -0,0 +1,314 @@
+        gl.enableVertexAttribArray(coord);
+    }
+
+    function prepareProgram(gl, videoFrame, videoFrameHandle, samplerName) {
+        let newSignature = generateSignature(videoFrameHandle);
+        if (program !== null && newSignature === signature)
+            return;
+        signature = newSignature;
+
+        const vSource = `
+    attribute vec2 aVertexPosition;
+    attribute vec2 aTexCoord;
+    uniform vec4 uTexTransform;
+    varying mediump vec2 vTexCoord;
+    void main(void) {
+      gl_Position = vec4(aVertexPosition, 0.0, 1.0);
+      vTexCoord = aTexCoord * uTexTransform.zw + uTexTransform.xy;
+    }
+  `;
+
+
+        let fSource =
+            emitRequiredExtension(videoFrameHandle) +
+            "precision mediump float;\n" +
+            "varying mediump vec2 vTexCoord;\n" +
+            emitSamplerDeclaration(videoFrameHandle, samplerName) +
+            videoFrameHandle.colorConversionShaderFunc +
+            "void main() {\n" +
+            emitSamplerBody(videoFrameHandle, samplerName, "vTexCoord", "pixel") +
+            emitUnPremultiplyAlpha(videoFrameHandle, "pixel") +
+            "  pixel.rgb = DoColorConversion(pixel.xyz);\n" +
+            emitPremultiplyAlpha(videoFrameHandle, "pixel") +
+            "  gl_FragColor = pixel;\n" +
+            "}\n";
+
+        program = initShaderProgram(gl, vSource, fSource);
+        initVertexBuffers(gl, program, videoFrame, "aVertexPosition", "aTexCoord");
+        let locTexTransform = gl.getUniformLocation(program, "uTexTransform");
+        let transform = new Float32Array([0, 0, 1, 1]);
+        if (videoFrameHandle.flipY) {
+            transform[1] = 1 - transform[1];
+            transform[3] = -transform[3];
+        }
+
+        // TODO: adjust visible rect.
+
+        if (videoFrameHandle.textureInfoArray[0].samplerType === "sampler2DRect") {
+            transform[0] *= videoFrame.codedWidth;
+            transform[2] *= videoFrame.codedWidth;
+            transform[1] *= videoFrame.codedHeight;
+            transform[3] *= videoFrame.codedHeight;
+        }
+        gl.useProgram(program);
+        gl.uniform4fv(locTexTransform, transform);
+    }
+
+    let ready_frames = [];
+    let underflow = true;
+    let time_base = 0;
+    let program = null;
+    let signature = null;
+
+    function delay(time_ms) {
+        return new Promise((resolve) => {
+            setTimeout(resolve, time_ms);
+        });
+    }
+
+    function calculateTimeTillNextFrame(timestamp) {
+        if (time_base == 0)
+            time_base = performance.now();
+        let media_time = performance.now() - time_base;
+        return Math.max(0, (timestamp / 1000) - media_time);
+    }
+
+    async function renderFrame() {
+        if (ready_frames.length == 0) {
+            underflow = true;
+            return;
+        }
+        let frame = ready_frames.shift();
+        underflow = false;
+
+        let videoFrameHandle = ext.importVideoFrame(frame);
```

Comment:
The implementation in https://chromium-review.googlesource.com/2525285 throws exceptions if the video frame is not configured as expected. Should this demo handle those cases?

Reply:
Done!